pkg/egresspolicy/egresspolicy.go

```diff
@@ -0,0 +1,566 @@
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "egresspolicy")
+)
+
+// EgressPolicyConfig is the internal representation of Cilium Egress NAT Policy.
+type EgressPolicyConfig struct {
+	// id is the parsed config name and namespace
+	id types.NamespacedName
+	// uid is the unique identifier assigned by Kubernetes
+	uid types.UID
+
+	endpointSelectors []api.EndpointSelector
+	dstCIDRs          []*net.IPNet
+	egressIP          net.IP
+}
+
+// The egresspolicy manager stores the internal data tracking the policy
+// and pod mappings. And also hooks up all the callbacks to update the egress
+// map accordingly.
+type Manager struct {
+	mutex lock.Mutex
+
+	// Stores pod to policy mapping
+	policyPods map[podID][]policyID
+	// Stores policy configs indexed by policyID
+	policyConfigs map[policyID]*EgressPolicyConfig
+	// Stores podId to pod metadata mapping
+	podDataStore map[podID]*podMetadata
+
+	storeGetter StoreGetter
+}
+
+// podMetadata stores relevant metadata associated with a pod that's updated during pod
+// add/update events
+type podMetadata struct {
+	// Pod labels
+	labels map[string]string
+	// Namespace labels
+	nsLabels map[string]string
+	// id the pod's name and namespace
+	id podID
+	// ips are pod's unique IPs
+	ips []string
+}
+
+// podID is pod name and namespace
+type podID = types.NamespacedName
+
+// PolicyID includes policy name and namespace
+type policyID = types.NamespacedName
+
+type StoreGetter interface {
+	GetStore(name string) cache.Store
+}
+
+// policyConfigSelectsPod determines if the given pod is selected by the policy
+// config based on matching labels of config and pod.
+func (config *EgressPolicyConfig) policyConfigSelectsPod(podInfo *podMetadata) bool {
+	labelsToMatch := labels.Set(podInfo.labels)
+	nsLabels := labels.Set(podInfo.nsLabels)
+	for k, v := range nsLabels {
+		labelsToMatch[policy.JoinPath(k8sConst.PodNamespaceMetaLabels, k)] = v
+	}
+	for _, selector := range config.endpointSelectors {
+		if selector.Matches(labelsToMatch) {
+			return true
+		}
+	}
+	return false
+}
+
+func Parse(cenp *v2alpha1.CiliumEgressNATPolicy) (*EgressPolicyConfig, error) {
+	var endpointSelectorList []api.EndpointSelector
+	var dstCidrList []*net.IPNet
+
+	allowAllNamespacesRequirement := slim_metav1.LabelSelectorRequirement{
+		Key:      k8sConst.PodNamespaceLabel,
+		Operator: slim_metav1.LabelSelectorOpExists,
+	}
+	name := cenp.ObjectMeta.Name
+
+	if name == "" {
+		return nil, fmt.Errorf("CiliumEgressNATPolicy must have a name")
+	}
+
+	for _, cidrString := range cenp.Spec.DestinationCIDRs {
+		_, cidr, err := net.ParseCIDR(cidrString)
+		if err != nil {
+			log.WithError(err).Warnf("CiliumEgressNATPolicy %v error parsing cidr", name)
+			continue
+		}
+		dstCidrList = append(dstCidrList, cidr)
+	}
```

Comment:
Shouldn't this be done upfront by the apiserver from the validation defined on the CRD field? Do we need to validate it again? Also, why do we warn if we fail to parse a CIDR? Shouldn't we return an error here? We shouldn't operate on half bad data.

Reply:
I added validation pattern for it in the type definition and now it is validated by the api server.
$ kubectl apply -f sample_tmp.yaml
The CiliumEgressNATPolicy "egress-sample" is invalid: spec.destinationCidrs: Invalid value: "8dsfsdewr": spec.destinationCidrs in body should match '^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/([0-9]|[1-2][0-9]|3[0-2])$'

However this err is there just for the correctness. The checking done by net is more sophisticated than the kubebuilder, so I decided to leave it here. And I changed to early return if we encounter a parse error, which I though is a good suggestion.

The change unfortunately is more extensive since it updates the crd part of the code too.
thx!